From ed29277f146b113dc71e1762fcc5715101e07ab8 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Mon, 11 Jun 2018 17:50:13 +0530
Subject: [PATCH 0358/1851] drm: xlnx: bridge: Add support for VTC as bridge

commit 7e26f94acee4969ce572e56c843310201cd8c68f from
https://github.com/Xilinx/linux-xlnx.git

Adds support to have VTC as a bridge by adding API for
setting timings based on the videomode.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_bridge.c | 29 +++++++++++++++++++++++++++++
 drivers/gpu/drm/xlnx/xlnx_bridge.h | 13 +++++++++++++
 2 files changed, 42 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_bridge.c b/drivers/gpu/drm/xlnx/xlnx_bridge.c
index ad706a5666bf..6ee462ada676 100644
--- a/drivers/gpu/drm/xlnx/xlnx_bridge.c
+++ b/drivers/gpu/drm/xlnx/xlnx_bridge.c
@@ -52,6 +52,7 @@ struct xlnx_bridge_helper {
 
 static struct xlnx_bridge_helper helper;
 
+struct videomode;
 /*
  * Client functions
  */
@@ -209,6 +210,34 @@ int xlnx_bridge_get_output_fmts(struct xlnx_bridge *bridge,
 }
 EXPORT_SYMBOL(xlnx_bridge_get_output_fmts);
 
+/**
+ * xlnx_bridge_set_timing - Set the video timing
+ * @bridge: bridge to set
+ * @vm: Videomode
+ *
+ * Set the video mode so that timing can be generated using this
+ * by the video timing controller.
+ *
+ * Return: 0 on success. -ENOENT if no callback, -EFAULT if in error state,
+ * or return code from callback.
+ */
+int xlnx_bridge_set_timing(struct xlnx_bridge *bridge, struct videomode *vm)
+{
+	if (!bridge)
+		return 0;
+
+	if (helper.error)
+		return -EFAULT;
+
+	if (bridge->set_timing) {
+		bridge->set_timing(bridge, vm);
+		return 0;
+	}
+
+	return -ENOENT;
+}
+EXPORT_SYMBOL(xlnx_bridge_set_timing);
+
 /**
  * of_xlnx_bridge_get - Get the corresponding Xlnx bridge instance
  * @bridge_np: The device node of the bridge device
diff --git a/drivers/gpu/drm/xlnx/xlnx_bridge.h b/drivers/gpu/drm/xlnx/xlnx_bridge.h
index 2dd5ca196a9a..64330169bd22 100644
--- a/drivers/gpu/drm/xlnx/xlnx_bridge.h
+++ b/drivers/gpu/drm/xlnx/xlnx_bridge.h
@@ -19,6 +19,8 @@
 #ifndef _XLNX_BRIDGE_H_
 #define _XLNX_BRIDGE_H_
 
+struct videomode;
+
 struct xlnx_bridge_debugfs_file;
 
 /**
@@ -32,6 +34,7 @@ struct xlnx_bridge_debugfs_file;
  * @get_input_fmts: callback to get supported input formats.
  * @set_output: callback to set the output
  * @get_output_fmts: callback to get supported output formats.
+ * @set_timing: callback to set timing in connected video timing controller.
  * @debugfs_file: for debugfs support
  */
 struct xlnx_bridge {
@@ -48,6 +51,7 @@ struct xlnx_bridge {
 			  u32 width, u32 height, u32 bus_fmt);
 	int (*get_output_fmts)(struct xlnx_bridge *bridge,
 			       const u32 **fmts, u32 *count);
+	int (*set_timing)(struct xlnx_bridge *bridge, struct videomode *vm);
 	struct xlnx_bridge_debugfs_file *debugfs_file;
 };
 
@@ -75,6 +79,7 @@ int xlnx_bridge_set_output(struct xlnx_bridge *bridge,
 			   u32 width, u32 height, u32 bus_fmt);
 int xlnx_bridge_get_output_fmts(struct xlnx_bridge *bridge,
 				const u32 **fmts, u32 *count);
+int xlnx_bridge_set_timing(struct xlnx_bridge *bridge, struct videomode *vm);
 struct xlnx_bridge *of_xlnx_bridge_get(struct device_node *bridge_np);
 void of_xlnx_bridge_put(struct xlnx_bridge *bridge);
 
@@ -141,6 +146,14 @@ static inline int xlnx_bridge_get_output_fmts(struct xlnx_bridge *bridge,
 	return 0;
 }
 
+static int xlnx_bridge_set_timing(struct xlnx_bridge *bridge,
+				  struct videomode *vm)
+{
+	if (bridge)
+		return -ENODEV;
+	return 0;
+}
+
 static inline struct xlnx_bridge *
 of_xlnx_bridge_get(struct device_node *bridge_np)
 {
-- 
2.31.1

